### PR TITLE
DC/DEBUG: added dci alloc assert, allow missorder

### PR DESCRIPTION
**Pull request by hoopoepg**
- allow missordered dci allocation in some tests
- some tests update resources manually to make some corner cases in UCX, due to this dci allocation ordering may be broken. suppress assertion in such cases

### CODE REVIEW

File: src/uct/ib/dc/dc_mlx5.c [hunk 1]
```diff
@@ -1334,6 +1334,10 @@ static UCS_CLASS_INIT_FUNC(uct_dc_mlx5_iface_t, uct_md_h tl_md, uct_worker_h wor
     self->flags                            = 0;
     kh_init_inplace(uct_dc_mlx5_fc_hash, &self->tx.fc_hash);
 
+#if UCS_ENABLE_ASSERT
+    self->tx.allow_misordering             = 0;
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5.c:1338]: it can be dc_iface flag (in self->flags) then no need special debug-mode handling

File: test/gtest/uct/ib/test_dc.cc [hunk 2]
```diff
@@ -452,6 +452,28 @@ UCT_DC_INSTANTIATE_TEST_CASE(test_dc)
...
+            m_iface->tx.allow_misordering = 1;
+#endif
+        }
+
+        ~allow_misordering()
+        {
+#if UCS_ENABLE_ASSERT
+            m_iface->tx.allow_misordering = 0;
+#endif
+        }
```
Comment (yosefe) [inline test/gtest/uct/ib/test_dc.cc:471]: IMO overcomplication - just set the allow flag once during the test, no need to remove this flag

File: src/uct/ib/dc/dc_mlx5_ep.h [hunk 3]
```diff
@@ -554,10 +554,8 @@ uct_dc_mlx5_iface_dci_get(uct_dc_mlx5_iface_t *iface, uct_dc_mlx5_ep_t *ep)
     }
 
     if (uct_dc_mlx5_iface_dci_can_alloc(iface, pool_index)) {
-        /* TODO: disabled due to shot on some tests where resources are updated
-         * manually: dc_mlx5/test_dc_flow_control.pending_grant for instance. */
-        /* TODO: look how to eliminate assert on such tests */
-        /* ucs_assert(ucs_arbiter_is_empty(waitq)); */
+        waitq = uct_dc_mlx5_iface_dci_waitq(iface, pool_index);
+        ucs_assert(iface->tx.allow_misordering || ucs_arbiter_is_empty(waitq));
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.h:558]:
```
if (!(self->flags & IGNORE_DCI_WAITQ_REORDER)) {
    waitq = uct_dc_mlx5_iface_dci_waitq(iface, pool_index);
    ucs_assert(ucs_arbiter_is_empty(waitq));
}
```

File: test/gtest/uct/ib/test_dc.cc [hunk 4]
```diff
@@ -517,6 +539,8 @@ UCS_TEST_P(test_dc_flow_control, general_disabled)
 
 UCS_TEST_P(test_dc_flow_control, pending_grant)
 {
+    allow_misordering misorder(m_e2);
```
Comment (yosefe) [inline test/gtest/uct/ib/test_dc.cc:542]: pls comment in each test why waitq reorder happens

File: src/uct/ib/dc/dc_mlx5.c (hunk 1 above)
Comment (hoopoepg) [reply src/uct/ib/dc/dc_mlx5.c]: > it can be dc_iface flag (in self->flags) then no need special debug-mode handling

used flags

File: src/uct/ib/dc/dc_mlx5_ep.h (hunk 3 above)
Comment (hoopoepg) [reply src/uct/ib/dc/dc_mlx5_ep.h]: > if (!(self->flags & IGNORE_DCI_WAITQ_REORDER)) {
>     waitq = uct_dc_mlx5_iface_dci_waitq(iface, pool_index);
>     ucs_assert(ucs_arbiter_is_empty(waitq));
> }

done

File: test/gtest/uct/ib/test_dc.cc (hunk 2 above)
Comment (hoopoepg) [reply test/gtest/uct/ib/test_dc.cc]: > IMO overcomplication - just set the allow flag once during the test, no need to remove this flag

ok, used single routine

File: test/gtest/uct/ib/test_dc.cc (hunk 4 above)
Comment (hoopoepg) [reply test/gtest/uct/ib/test_dc.cc]: > pls comment in each test why waitq reorder happens

added comment

**Comment by hoopoepg** [comment]
@yosefe ok to merge?

**Review by brminich — APPROVED** [review]
